shell/platform/android/surface_texture_external_texture_gl.cc

```diff
@@ -0,0 +1,115 @@
+#include "third_party/skia/include/core/SkAlphaType.h"
+#include "third_party/skia/include/core/SkColorSpace.h"
+#include "third_party/skia/include/core/SkColorType.h"
+#include "third_party/skia/include/core/SkImage.h"
+#include "third_party/skia/include/gpu/GrBackendSurface.h"
+#include "third_party/skia/include/gpu/GrDirectContext.h"
+#include "third_party/skia/include/gpu/ganesh/SkImageGanesh.h"
+#include "third_party/skia/include/gpu/ganesh/gl/GrGLBackendSurface.h"
+#include "third_party/skia/include/gpu/gl/GrGLTypes.h"
+
+namespace flutter {
+
+SurfaceTextureExternalTextureGL::SurfaceTextureExternalTextureGL(
+    int64_t id,
+    const fml::jni::ScopedJavaGlobalRef<jobject>& surface_texture,
+    std::shared_ptr<PlatformViewAndroidJNI> jni_facade)
+    : SurfaceTextureExternalTexture(id, surface_texture, jni_facade) {}
+
+SurfaceTextureExternalTextureGL::~SurfaceTextureExternalTextureGL() {
+  Detach();
+}
+
+void SurfaceTextureExternalTextureGL::ProcessFrame(PaintContext& context,
+                                                   const SkRect& bounds) {
+  if (state_ == AttachmentState::kUninitialized) {
+    // Generate the texture handle.
+    glGenTextures(1, &texture_name_);
+    Attach(texture_name_);
+  }
+  FML_CHECK(state_ == AttachmentState::kAttached);
+
+  // Updates the texture contents and transformation matrix.
+  Update();
+
+  // Create a
+  GrGLTextureInfo textureInfo = {GL_TEXTURE_EXTERNAL_OES, texture_name_,
+                                 GL_RGBA8_OES};
+  auto backendTexture =
+      GrBackendTextures::MakeGL(1, 1, skgpu::Mipmapped::kNo, textureInfo);
+  dl_image_ = DlImage::Make(SkImages::BorrowTextureFrom(
+      context.gr_context, backendTexture, kTopLeft_GrSurfaceOrigin,
+      kRGBA_8888_SkColorType, kPremul_SkAlphaType, nullptr));
+}
+
+void SurfaceTextureExternalTextureGL::Detach() {
+  SurfaceTextureExternalTexture::Detach();
+  if (texture_name_ != 0) {
+    glDeleteTextures(1, &texture_name_);
+    texture_name_ = 0;
+  }
+}
+
+SurfaceTextureExternalTextureImpellerGL::
+    SurfaceTextureExternalTextureImpellerGL(
+        const std::shared_ptr<impeller::ContextGLES>& context,
+        int64_t id,
+        const fml::jni::ScopedJavaGlobalRef<jobject>& surface_texture,
+        std::shared_ptr<PlatformViewAndroidJNI> jni_facade)
+    : SurfaceTextureExternalTexture(id, surface_texture, jni_facade),
+      impeller_context_(context) {}
+
+SurfaceTextureExternalTextureImpellerGL::
+    ~SurfaceTextureExternalTextureImpellerGL() {
+  Detach();
+}
+
+void SurfaceTextureExternalTextureImpellerGL::ProcessFrame(
+    PaintContext& context,
+    const SkRect& bounds) {
+  if (state_ == AttachmentState::kUninitialized) {
+    // Generate the texture handle.
+    impeller::TextureDescriptor desc;
+    desc.type = impeller::TextureType::kTextureExternalOES;
+    desc.storage_mode = impeller::StorageMode::kDevicePrivate;
+    desc.format = impeller::PixelFormat::kR8G8B8A8UNormInt;
```

Comment:
Same comment as last PR: we need to look this up at runtime right?

Reply:
Interesting! right, I guess its GL. I think we also do something similar in the GLES Impeller backend.